File: EstimateSummation.py
from mpi4py import MPI
import csv
import sys
import time

# read csv data
def read_csv(filename):
    data = []
    with open(filename, 'r') as file:
        csv_reader = csv.DictReader(file)
        for row in csv_reader:
            data.append(row)
    return data

filename = "../../migration.csv"

# create connection to file
comm = MPI.COMM_WORLD
size = comm.Get_size()
rank = comm.Get_rank()

if rank == 0:
    data = read_csv(filename)
    chunks = [[] for _ in range(size)]
    for i, row in enumerate(data):
        chunks[i % size].append(row)
else:
    chunks = None

# start timer
start_time = time.time()
# Scatter chunks to all processes
local_chunk = comm.scatter(chunks, root=0)
print("Process", rank, "received", len(local_chunk), "rows of data")

# Perform local computation (sum of 'estimate' column in the chunk)
local_sum = sum(int(row['estimate']) for row in local_chunk)

# Gather local sums to root process
global_sums = comm.gather(local_sum, root=0)

# Compute total sum in root process
if rank == 0:
    total_sum = sum(global_sums)
    # end timer
    end_time = time.time()

    print("Sum of 'estimate' column:", total_sum)
    print("Time taken for sum calculation: ", (end_time - start_time) * 1000, " milliseconds")

File: EstimateSummation.java
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

import java.util.Arrays;
import java.util.List;

import static org.apache.spark.sql.functions.sum;

public class EstimateSummation {
    public static void main(String[] args) {
        // Create a SparkSession
        SparkSession spark = SparkSession
                .builder()
                .getOrCreate();

        // Create a JavaSparkContext
        JavaSparkContext sc = new JavaSparkContext(spark.sparkContext());

        // Define the schema
        String schemaString = "year_month,month_of_release,passenger_type,direction,citizenship,visa,country_of_residence,estimate,standard_error,status";
        String[] fields = schemaString.split(",");

        // Convert the fields array into a list
        List<String> fieldList = Arrays.asList(fields);

        StructField[] structFields = sc.parallelize(fieldList).map(fieldName ->
                DataTypes.createStructField(fieldName, DataTypes.StringType, true)
        ).collect().toArray(new StructField[0]);

        StructType schema = DataTypes.createStructType(structFields);

        // Read CSV
        Dataset<Row> sdfData = spark.read()
                .format("csv")
                .option("header", "true")
                .schema(schema)
                .load("../../migration.csv");

        // Explicitly repartition the DataFrame
        sdfData = sdfData.repartition(3);

        // Start timer
        long startTime = System.currentTimeMillis();

        // Sum all of the column Age values in parallel
        Dataset<Row> sumAge = sdfData.agg(sum(sdfData.col("estimate")));
        sumAge.show();

        // End timer
        long endTime = System.currentTimeMillis();
        long elapsedTime = endTime - startTime;
        System.out.println("Time taken for sum calculation: " + elapsedTime + " milliseconds");

        spark.stop();
    }
}
